libkbfs/tlf_journal.go

```diff
@@ -375,6 +376,134 @@ func (j *tlfJournal) resumeBackgroundWork() {
 	}
 }
 
+func (j *tlfJournal) getBlockRangeToFlush(ctx context.Context) (
+	doFlush bool, first journalOrdinal, last journalOrdinal, err error) {
+	j.journalLock.RLock()
+	defer j.journalLock.RUnlock()
+
+	length, err := j.blockJournal.length()
+	if err != nil {
+		return false, 0, 0, err
+	}
+	if length == 0 {
+		return false, 0, 0, nil
+	}
+
+	first, last, err = j.blockJournal.ordinalRange()
+	if err != nil {
+		return false, 0, 0, err
+	}
+	return true, first, last, nil
+}
+
+func (j *tlfJournal) getBlockEntryToFlush(ctx context.Context,
+	ordinal journalOrdinal) (*blockJournalEntry, []byte,
+	BlockCryptKeyServerHalf, error) {
+	j.journalLock.RLock()
+	defer j.journalLock.RUnlock()
+	return j.blockJournal.getEntryToFlush(ctx, ordinal)
+}
+
+func (j *tlfJournal) removeFlushedBlockEntry(ctx context.Context,
+	ordinal journalOrdinal, entry blockJournalEntry) (int64, error) {
+	j.journalLock.Lock()
+	defer j.journalLock.Unlock()
+	return j.blockJournal.removeFlushedEntry(ctx, ordinal, entry)
+}
+
+func (j *tlfJournal) flushBlocks(ctx context.Context,
+	first journalOrdinal, last journalOrdinal) (err error) {
+	j.flushLock.Lock()
+	defer j.flushLock.Unlock()
+
+	j.log.CDebugf(ctx, "Flushing block entries %d to %d", first, last)
+	defer func() {
+		j.log.CDebugf(ctx, "Finished flushing block entries: %v", err)
+	}()
+
+	putBps := newBlockPutState(int(last-first) + 1)
+	addRefBps := newBlockPutState(int(last-first) + 1)
+	var allEntries []*blockJournalEntry
+	var otherEntries []*blockJournalEntry
+
+	for ordinal := first; ordinal <= last; ordinal++ {
```

Comment:
Yeah, I think it would be better to delegate the responsibility of figuring out which blocks ops are parallelizable to `blockJournal`. Concretely:

1) Change `blockJournal.getNextEntryToFlush` to `getNextEntriesToFlush`, which returns a list of structs representing ops that can be done in parallel, where each struct will contain whatever `getNextEntryToFlush` is returning now. Then `blockJournal` will return either a sequential run of `put` ops, or just a single op.

2) Change `blockJournal.removeFlushedBlockEntry` to `removeFlushedBlockEntries` and to take a list, and have it check that the given list matches the head entries of the journal. To answer your question in one of my pending PRs, it's not strictly necessary to pass it the entire list, since it just needs to now how many entries to drop from the front, but I'm just doing the comparison to guard against race conditions, since only one thing should be flushing at a time.

3) change `flushBlockJournalEntry` to `flushBlockJournalEntries`, and have it peek into the list, and if it contains only puts, it can do them in parallel, and otherwise it just does them sequentially.

4) Change the plumbing from `tlfJournal` as necessary.

Looks like that's close to what you have now. What do you think?